python/pyarrow/_dataset.pyx

```diff
@@ -633,7 +633,7 @@ cdef class FileSystemDataSource(DataSource):
         CFileSystemDataSource* filesystem_source
 
     def __init__(self, Schema schema not None,
-                 Expression source_partition not None,
+                 Expression source_partition,
```

Comment:
You'll need to properly handle the case where `source_partition == None`; IIUC this will pass a nullptr to C++. The default should be `scalar(true)`

Reply:
@bkietz hmm, there are tests that pass None, and they pass (but not sure if they cover the behaviour that would trigger problems in C++)

Reply:
Correct: there are python tests which directly construct a source with null source_partition, but those don't scan the source and so the issue won't arise

Reply:
Indeed, for all other tests we actually use the Factory and never construct a Source directly. Will change.